libcxx/include/cmath

```diff
@@ -765,6 +766,53 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp __constexpr_scalbn(_Tp _
   return __builtin_scalbn(__x, __exp);
 }
 
+#if _LIBCPP_STD_VER >= 17
+
+/// \return the Hermite polynomial \f$ H_n(x) \f$
+/// \note The implementation is based on the recurrence formula
+/// \f[
+///   H_{n+1}(x) = 2x H_n(x) - 2n H_{n-1}
+/// \f]
+/// Press, William H., et al. Numerical recipes 3rd edition: The art of
+/// scientific computing. Cambridge university press, 2007, p. 183.
+template <class _Real>
+_LIBCPP_HIDE_FROM_ABI _Real __hermite(unsigned __n, _Real __x) {
+  if (std::isnan(__x))
+    return std::numeric_limits<_Real>::quiet_NaN();
+
+  _Real __H_0{1};
+  if (__n == 0)
+    return __H_0;
+
+  _Real __H_n_prev = __H_0;
+  _Real __H_n      = 2 * __x;
+  for (unsigned __i = 1; __i < __n; ++__i) {
+    _Real __H_n_next = 2 * (__x * __H_n - __i * __H_n_prev);
+    __H_n_prev       = __H_n;
+    __H_n            = __H_n_next;
+  }
+  return __H_n;
```

Comment:
I think we could have somewhat accurate overflow detection here with:
```
if (std::isnan(__H_n)) {
  // Overflow.
  if (n & 1)
    // n is odd, return infinity of the same sign as x.
    return std::copysign(std::numeric_limits<_Real>::infinity(), x);
  // n is even, return +Inf
  return std::numeric_limits<_real>::infinity();
}
return __H_n;
```

Reply:
While I was thinking about your comment for quite some time multiple issues crossed my mind.

* I assume your point is that in the nature of subtraction and recursive calls (e.g. `_Real __H_n_next = 2 * (__x * __H_n - __i * __H_n_prev);`) one could produce an Inf of the wrong sign?
  * I will try to find a test set which would show this issue for the current implementation.
* I believe one needs to clearly separate between NaNs and Infs.
  * NaNs are already handled at function entry. They should not be producible within this function.
  * The check in your code snippet should probably test for Infs, i.e. `if (std::isinf(__H_n)) { ... }`?
* This does clutter the implementation quite a bit (mixing error checks with formula). I assume similar logic will need to apply for some of the other special math functions. We might probably want to extract that logic at some point.

Reply:
+1 to everything in your comment.
- For the first point a simple non-trivial (might be not that useful) example is when `x = std::numeric_limits<_Real>::max()`, following the iterations, you will get that:
```
  hermite(0, x) = 1.0
  hermite(1, x) = +Inf
  hermite(2, x) = +Inf
  hermite(3, x) = NaN
  ...
```
and there are definitely more intricate cases.
- Another issue that we might want to address later is precision loss.  I haven't looked into the stability of the iterations, but we might have catastrophic cancellation when `__x * __H_n ~ __i * __H_n_prev`.  It is interesting to find some non-trivial examples, but we can defer dealing with this case to later PR.

Reply:
We could make use of your proposed code snippet at the entry of the function, e.g.
```cpp
_LIBCPP_HIDE_FROM_ABI _Real __hermite(unsigned __n, _Real __x) {
  if (std::isnan(__x))
    return __x;
  if (std::isinf(__x)) {
    // ... your proposed overflow handling
  }
  // ...
```
Though I am unsure of the importance of handling `std::hermite(n, +-Inf)` correctly. Personally I would not expect correct output for such an input (informally known as "garbage in, garbage out" ;D).
Let me know if this is worth implementing.

Reply:
I'm thinking of something like this:
https://godbolt.org/z/7MaYYzob9

Reply:
Looks good. Can I ask why you check for `std::isnan` instead of `std::isinf`? I believe I have seen examples where at some `n` the next 2 results were `Inf` and then turned to `NaN`. If I find some more time I will try to find such an example.

Reply:
So I think for finite `__x`, if some intermediate `__H_n` is `Inf`, it is "correctly" overflowed (I might need to prove it or if it is not right, find some counter example).  So checking `isnan` should be enough.  OTOH, probably we should just use `!std::isfinite(__H_n)` instead.

Reply:
I looked some more into possible overflows. First thing to notice is that double overflows at ~1e308.
Hermite polynomial H_127 does not reach such large values inbetween any roots but only for larger x:
* largest root is at x=~16
* function values previous to the last root are significantly below the overflow threshold, i.e. |H_127(x)| < ~1e190 for |x| < ~16
* overflow threshold is reached at roughly: H_127(x=140)~1e308

For polynomials of smaller order it is even more extreme (e.g. largest root is smaller and overflow happens at even larger x).
I would not take polynomials of higher order into this discussion as their result is deemed implementation-defined by the Standard. Thus, possibly slowing down code for n<128 because of arguments for n>=128 seems strange.

My conclusion: I believe your initial suggestion is enough (e.g. checking the value of `__H_n` *after* the loop). Using `!std::isfinite()` also seems the approriate check here.

I will probably push a change similar to `hermite2` as in this [godbolt](https://godbolt.org/z/6xojME18W). (`hermite` is the current implementation, and `hermite1` the one you proposed in the previous godbolt.)

Reply:
Implemented in 511b45679ab65dc3efca81abc976c27711a5df9a.

Reply:
SGTM